src/serde/tree_cache.rs

```diff
@@ -0,0 +1,744 @@
+    /// is the sentinel node, which means we can't compute the tree hash.
+    tree_hash: Option<Bytes32>,
+    /// a node can have an arbitrary number of parents, since they can be reused
+    /// this is a list of parent nodes, followed by whether we're the left or
+    /// right child. The u32 is an index into the node_entry vector.
+    parents: Vec<(u32, ChildPos)>,
+    /// if this node doesn't have a tree_hash, the serialized length is not
+    /// valid as it cannot be computed.
+    serialized_length: u64,
+    /// set to non-zero, if this node is pushed onto the parse stack. Since a
+    /// node can be on the stack multiple times, it's counted up every time it's
+    /// pushed, and counted down every time it's popped.
+    pub on_stack: u32,
+}
+
+struct PartialPath {
+    // the path we've built so far
+    path: PathBuilder,
+    // if we're traversing the stack, this is the stack position. Note that this
+    // is not an index into the stack array, it's a counter of how far away from
+    // the top of the stack we are. 0 means we're at the top, and we've found
+    // a path.
+    // -1 means we're not traversing the stack.
+    // TODO: these fields are mutually exclusive (stack_pos vs. idx and child).
+    // it might be nice to use an enum, but it's not obvious that it's simpler
+    // or faster
+    stack_pos: i32,
+    // if we're not traversing the stack, this is the next parent, and whether
+    // we're coming from the left or right child
+    idx: u32,
+    child: ChildPos,
+}
+
+enum CacheOp {
+    // node to traverse
+    Traverse(NodePtr),
+    // pair node
+    Cons(NodePtr),
+}
+
+#[derive(Clone)]
+pub struct TreeUndoState {
+    stack: Vec<u32>,
+    lookup: VisitedNodes,
+    sentinel_entry: Option<u32>,
+}
+
+#[derive(Default)]
+pub struct TreeCache {
+    /// caches extra metadata about a tree of nodes. The value is an index into
+    /// the node_entry vector.
+    cache: HashMap<NodePtr, u32>,
+
+    /// The metadata for all nodes in the tree. This is like a shadow tree
+    /// structure to the NodePtr one. The most important difference is that
+    /// identical nodes are merged, using the same NodeEntry, and additional
+    /// metadata is kept, such as the tree hash.
+    node_entry: Vec<NodeEntry>,
+
+    /// maps tree-hashes to the index of the corresponding NodeEntry in the
+    /// node_entry vector. For any given tree hash, we're only supposed to
+    /// have a single NodeEntry. There may be multiple NodePtr referring to
+    /// the same NodeEntry (if they are identical sub trees).
+    hash_to_node: HashMap<Bytes32, u32, RandomState>,
+
+    /// When deserializing, we keep a stack of tokens we've parsed so far, this
+    /// stack is maintaining that same state, since that's what back-references
+    /// are pointing into. Nodes that have the sentinel as direct decendants
+    /// don't have a tree-hash, so it's set to None for those entries
```

Comment:
When you say `None`, do you mean `0`?

Reply:
no, that sentence is referring to the `tree_hash` field in `NodeEntry`. I'm not sure why I thought that comment belonged here.